### PR TITLE
07778 Enhance tracking of reason for state save

### LINKED ISSUE #7778: Improve reporting of the reason a state is being written to disk
Currently, even states written due to a reconnect are being reported as a `periodic snapshot`

Add an enum with all potential reasons that a state might be saved to disk, and use it throughout in order to have an accurate report of why a given state is being written

### PR DESCRIPTION
**Pull request by alittley**
- Closes #7778

### CODE REVIEW

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/components/state/DefaultStateManagementComponent.java [hunk 1]
```diff
@@ -290,9 +292,14 @@ public DefaultStateManagementComponent(
      * @param signedState the newly complete signed state
      */
     private void stateHasEnoughSignatures(final SignedState signedState) {
-        if (signedState.isStateToSave()) {
-            signedStateFileManager.saveSignedStateToDisk(signedState);
+        final StateToDiskReason reason = signedState.getStateToDiskReason();
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/components/state/DefaultStateManagementComponent.java:295]: Even if you don't store a boolean value internally in the `SignedState` object, I think the method `isStateToSave()` makes this code more readable. Triggering state saving whenever reason is non-null seems like a funky API.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java [hunk 2]
```diff
@@ -487,21 +488,22 @@ public long getMinRoundGeneration() {
...
 
     /**
-     * Check if this is a state that needs to be eventually written to disk.
+     * Set the reason why this state needs to be written to disk.
      *
-     * @return true if this state eventually needs to be written to disk
+     * @param reason the reason why this state needs to be written to disk
      */
-    public boolean isStateToSave() {
-        return stateToSave;
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java:495]: I recommend keeping this method. You can just do `return reason != null`.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java [hunk 3]
```diff
@@ -487,21 +488,22 @@ public long getMinRoundGeneration() {
...
+    public void setStateToDiskReason(@NonNull final StateToDiskReason reason) {
+        stateToDiskReason = reason;
     }
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java:497]: Perhaps for a clearer API, we should so something like this. That way it is clear that a side effect of this method is that this state will be written to disk.

```suggestion
    public void markAsStateToSave(@NonNull final StateToDiskReason reason) {
        stateToDiskReason = reason;
    }
```

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StateToDiskReason.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.Objects;
+
+/**
+ * The reason for writing the state to disk
+ */
+public enum StateToDiskReason {
+    /**
+     * The state was written to disk because the platform is starting up
+     */
+    FIRST_ROUND("first-round"),
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StateToDiskReason.java:29]: How about `FIRST_ROUND_AFTER_GENESIS`?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StateToDiskReason.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+package com.swirlds.platform.state.signed;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.util.Objects;
+
+/**
+ * The reason for writing the state to disk
+ */
+public enum StateToDiskReason {
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StateToDiskReason.java:25]: I would be ok if we just used the enum string, as opposed to providing a lower case version. Nobody currently depends on these exact strings (I think).

Comment (alittley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StateToDiskReason.java:25]: To clarify, you mean it's ok to use the enum strings for the directories where ISS and fatal states are saved?

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/components/state/DefaultStateManagementComponent.java (hunk 1 above)
Comment (alittley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/components/state/DefaultStateManagementComponent.java]: done

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java (hunk 2 above)
Comment (alittley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java]: done

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java (hunk 3 above)
Comment (alittley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/SignedState.java]: done

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StateToDiskReason.java (hunk 4 above)
Comment (alittley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StateToDiskReason.java]: done

**Review by lpetrovic05 — APPROVED** [review]

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StateToDiskReason.java (hunk 5 above)
Comment (cody-littley) [reply platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StateToDiskReason.java:22]: No, we should probably keep the same file paths. I'm think JRS might be hard coded to them. I didn't realize these strings were also generating the folder structure, thought it was just a logging thing.

File: platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StateToDiskReason.java [hunk 6]
```diff
@@ -16,32 +16,60 @@
...
+    StateToDiskReason(@NonNull final String description) {
+        this.description = Objects.requireNonNull(description);
+    }
+
+    /**
+     * Get the description of the reason
+     *
+     * @return the description of the reason
+     */
+    public String getDescription() {
```
Comment (cody-littley) [inline platform-sdk/swirlds-platform-core/src/main/java/com/swirlds/platform/state/signed/StateToDiskReason.java:72]:
```suggestion
     */
     @NonNull
    public String getDescription() {
```

**Pull request merged**